subsys/bluetooth/host/uuid.c

```diff
@@ -105,10 +105,12 @@ void bt_uuid_to_str(const struct bt_uuid *uuid, char *str, size_t len)
 
 	switch (uuid->type) {
 	case BT_UUID_TYPE_16:
-		snprintk(str, len, "%04x", BT_UUID_16(uuid)->val);
+		snprintk(str, len, "%04x",
+			 sys_cpu_to_le16(BT_UUID_16(uuid)->val));
```

Comment:
This looks very strange to me, since normally you'd want to pass host (cpu) endianness data to printk whereas various protocol structs is where the protocol endianness is found. Are you sure that you don't want `sys_le16_to_cpu()` here instead, i.e. `BT_UUID_16(uuid)->val` would be expected to always be little-endian?

Reply:
Indeed, I only tested this with 128b UUIDs and might have made wrong assumptions. Will check and fix if necessary.

Reply:
> You are completely right, UUIDs are stored in LE

@fnde-ot how did you make that assessment? I'm looking at `include/bluetooth/uuid.h` and I don't see any indication of it. Instead, it seems like 32- and 16-bit UUIDs store the value in host endianness. The only exception seems to be `BT_UUID_128_ENCODE()` which generates a byte array with the individual components in little endian format. Perhaps your original issue was really only with 128-bit, and the 16/32 variants should be untouched?

Reply:
Btw, what we probably should fix is the documentation for the UUID structs to make it clear what the endianness of the "value" is so this kind of confusion doesn't arise again.

Another thing that comes to mind (but is not so urgent): BR/EDR SDP might be skipping any conversions and therefore actually keeping the protocol endianness all the time (big-endian).

Reply:
Aren't the 16 and 32 bit values already in host endianness?
```
#define BT_UUID_INIT_16(value)		\
{					\
	.uuid = { BT_UUID_TYPE_16 },	\
	.val = (value),			\
}
#define BT_UUID_DECLARE_16(value) \
	((struct bt_uuid *) ((struct bt_uuid_16[]) {BT_UUID_INIT_16(value)}))

#define BT_UUID_HTS_VAL 0x1809
#define BT_UUID_HTS \
	BT_UUID_DECLARE_16(BT_UUID_HTS_VAL)
```

Does this print correctly on big endian?
```
	char str[BT_UUID_STR_LEN];
	bt_uuid_to_str(BT_UUID_HTS, str, sizeof(str));
	shell_print(ctx_shell, "HTS UUID %s", str);
```

Reply:
@jhedberg I was looking at BT_UUID_16_ENCODE, BT_UUID_32_ENCODE and BT_UUID_128_ENCODE.

Comments in uuid.h are a bit confusing, as according to it we are supposed to create 16-bit, 32-bit and 128-bit UUIDs using:
```
BT_DATA_BYTES(BT_DATA_UUID16_ALL, BT_UUID_16_ENCODE(0x180a))
BT_DATA_BYTES(BT_DATA_UUID32_ALL, BT_UUID_32_ENCODE(0x180a01af))
BT_UUID_DECLARE_128(BT_UUID_128_ENCODE(0x6E400001, 0xB5A3, 0xF393, 0xE0A9, 0xE50E24DCCA9E))
```

When in practice, you should do the following:
```
BT_UUID_DECLARE_16(0x180a)
BT_UUID_DECLARE_32(0x180a01af)
BT_UUID_DECLARE_128(BT_UUID_128_ENCODE(0x6E400001, 0xB5A3, 0xF393, 0xE0A9, 0xE50E24DCCA9E))
```

So the BT_UUID_16_ENCODE and BT_UUID_32_ENCODE are only used to encode 16-bit and 32-bit UUIDs into 128-bit UUIDs to use for example in advertisements.

(So yes, 16 and 32 bit values are already in host endianness!)

Reply:
@fnde-ot BT_UUID_16_ENCODE, BT_UUID_32_ENCODE and BT_UUID_128_ENCODE all encodes Little Endian byte arrays.
But only BT_UUID_DECLARE_128 takes it as input.

So `BT_DATA_BYTES(BT_DATA_UUID16_ALL, BT_UUID_16_ENCODE(0x180a))` is correct, because the advertising data has to be in little endian.